File: Chapter 2/BA2/BA2b/MedianString.py
#!/usr/bin/env python3
"""
Rosalind Problem ID: BA2b
Title: Implement MotifEnumeration
URL: https://rosalind.info/problems/BA2b/

Description:
Finds a k-mer Pattern that minimizes d(Pattern, Dna) over all k-mers 'Pattern', 
where d(Pattern, Dna) is the minimum hamming distance between 'Pattern' and an k-mer in 'Dna',
a set of 't' sequences with 'n' nucleotides of length.  We call such a k-mer a median string for Dna.
This is a pretty expensive algorithm with O(4^k.n.k.t) complexity.

Author: Santiago Wilders Azara
Date: 2026
"""

import sys

def hamming_distance(s1, s2):
    return sum(a != b for a, b in zip(s1, s2))

def numberToPattern(index, k):
    bases = ['A', 'C', 'G', 'T']
    pattern = ""
    for _ in range(k):
        pattern = bases[index % 4] + pattern
        index //= 4
    return pattern

def median_string(k, dna_sequences):
    min_count = float('inf')
    median = ''
    
    for i in range(4**k):
        pattern_i = numberToPattern(i, k)
        curr_count = 0
        
        for seq in dna_sequences:
            min_hd = float('inf')
            for j in range(len(seq) - k + 1):
                hd = hamming_distance(pattern_i, seq[j: j+k])
                if hd < min_hd: min_hd = hd
            curr_count += min_hd
        
        if curr_count < min_count:
            min_count = curr_count
            median = pattern_i
    
    return median


def main():
    if len(sys.argv) < 2:
        print("Usage: python MedianString.py <input_file.txt>")
        return

    file_path = sys.argv[1]

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = [line.strip() for line in f.readlines() if line.strip()]
            
            data = (lines[0], # k
                    lines[1:len(lines)]) # DNA Strings

    except FileNotFoundError:
        print(f"Error: The file '{file_path}' was not found.")
    except Exception as e:
        print(f"Unexpected error: {e}")
    
    if data:
        k, dna_strings = data
        median = median_string(int(k), dna_strings)
        print(median)

if __name__ == "__main__":
    main()

File: Chapter 2/Infering MTB DosR Motif Profile/median_string.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <limits>
#include <algorithm>

// Complexity: O(k)
int hamming_distance(const std::string& a, const std::string& b) {
    int dist = 0;
    for (size_t i = 0; i < a.size(); ++i)
        dist += (a[i] != b[i]);
    return dist;
}

// Complexity: O(k)
std::string number_to_pattern(int index, int k) {
    const char bases[] = {'A', 'C', 'G', 'T'};
    std::string pattern(k, 'A');
    for (int i = k - 1; i >= 0; --i) {
        pattern[i] = bases[index % 4];
        index /= 4;
    }
    return pattern;
}

// Complexity: O(4^k · n · k · t) — with branch-and-bound pruning
std::string median_string(int k, const std::vector<std::string>& dna) {
    int t = dna.size();
    int min_dist = std::numeric_limits<int>::max();
    std::string median = "";
    int total_patterns = 1;
    for (int i = 0; i < k; ++i) total_patterns *= 4;

    for (int i = 0; i < total_patterns; ++i) {
        std::string pattern = number_to_pattern(i, k);
        int curr_dist = 0;

        for (int s = 0; s < t; ++s) {
            const std::string& seq = dna[s];
            int n = seq.size();
            int min_hd = k;

            for (int j = 0; j <= n - k; ++j) {
                int hd = hamming_distance(pattern, seq.substr(j, k));
                if (hd < min_hd) min_hd = hd;
                if (min_hd == 0) break;  // early exit
            }

            curr_dist += min_hd;
            if (curr_dist >= min_dist) goto next_pattern;  // branch-and-bound
        }

        if (curr_dist < min_dist) {
            min_dist = curr_dist;
            median = pattern;
        }

        next_pattern:;
    }
    return median;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: ./median_string <input_file>\n";
        return 1;
    }

    std::ifstream file(argv[1]);
    if (!file) {
        std::cerr << "Error: cannot open file\n";
        return 1;
    }

    int k;
    file >> k;
    file.ignore();

    std::vector<std::string> dna;
    std::string line;
    while (std::getline(file, line))
        if (!line.empty()) dna.push_back(line);

    std::cout << median_string(k, dna) << "\n";
    return 0;
}